### PR TITLE
[RFC] Worker API description.

### PR DESCRIPTION
**Pull request by 3Hren**

### CODE REVIEW

**Comment by 3Hren** [comment]
Any notices about this stuff?

**Comment by mechmind** [comment]
Unhandled exception in one handler should not crash a worker.

File: rfc/worker.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Event handlers is a functional objects, that handle incoming events, obviously. Depending on handler type its arguments may vary.
+
+### Common Event Handler
+
+The commonnest event handler is a functional object, that accepts sender and receiver pair and returns nothing.
+
+```
+worker.on("ping", [](sender<streaming_tag<std::string>> tx, receiver<streaming_tag<std::string>> rx){
+    tx.send<io::streaming::chunk>(rx.recv().get()).get();
```
Comment (andrusha97) [inline rfc/worker.md]: Are there two different threads for running the handlers and for IO? Otherwise `rx.recv().get()` will deadlock.

Comment (3Hren) [inline rfc/worker.md]: Yep.

Comment (3Hren) [inline rfc/worker.md]: IO thread is completely hidden from users.

Number of calling threads is configurable.

**Comment by 3Hren** [comment]
@mechmind why it shouldn't?

You've violated the contract, it's your fault - take a core dump and debug.

You can always wrap with try/catch all potential dangerous places as you do in your normal code, I suppose.

File: rfc/worker.md (hunk 1 above)
Comment (andrusha97) [inline rfc/worker.md]: Ok.

Comment (3Hren) [inline rfc/worker.md]: But I think I should write about it in the RFC to make it clear for everyone.
Will do it tomorrow.

File: rfc/worker.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+The following code should make it clear:
+
+```cpp
+worker.on<http>("http", [](sender<fresh> tx, receiver<fresh> fresh_rx){
+    // This is the place for initialization.
+    ...
+
+    http_request rq;
+    receiver<streaming_tag<std::string>> rx;
+    std::tie(rq, rx) = fresh_rx.recv().get();
```
Comment (andrusha97) [inline rfc/worker.md]: Is there any way to get rid of this step? Something like
```
[](http_sender rep, http_receiver req){
  auto url = req.url();
  auto headers = req.headers();
  req.read_body();
  req.read_body();
  req.read_body();
  ...
}
```

**Comment by mechmind** [comment]
@3Hren Legitimate handlers would crash all of sudden in that case. Especially nasty if you serve more than one application from one worker (and contract do not forbid that, right?)

**Comment by 3Hren** [comment]
@mechmind just catch all exceptions :)

Look on the other hand, if the Framework will catch and eat all exceptions, how should you debug your non-properly-working worker?

File: rfc/worker.md (hunk 2 above)
Comment (3Hren) [inline rfc/worker.md]: You can always write your own event handler (like HTTP event handler was written). So, yes, there is a way.

But why do you want to do this? I think compile-time guarantee, that you can read headers (method, version, etc.) only once is better.

**Comment by 3Hren** [comment]
Updated.

**Pull request merged**